Dockerfile

```diff
@@ -1,39 +1,55 @@
-ARG base_image=ruby:2.7.6
-FROM ${base_image}
-
-# Add yarn to apt sources
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
-RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
-
-RUN apt-get update -qq && apt-get upgrade -y
-RUN apt-get install -y build-essential nodejs yarn && apt-get clean
-
-# This image is only intended to be able to run this app in a production RAILS_ENV
-ENV RAILS_ENV production
-
-ENV APP_HOME /app
-ENV DATABASE_URL mysql2://root:root@mysql/signon
-ENV GOVUK_APP_NAME signon
-ENV PORT 3016
-ENV TEST_DATABASE_URL mysql2://root:root@mysql/signon_test
-ENV ASSETS_PREFIX /assets/signon
-
-RUN mkdir $APP_HOME
-
-WORKDIR $APP_HOME
-ADD Gemfile* .ruby-version $APP_HOME/
-RUN bundle config set deployment 'true'
-RUN bundle config set without 'development test'
-RUN bundle install --jobs 4
-
-ADD . $APP_HOME
-
-RUN GOVUK_APP_DOMAIN=www.gov.uk \
-  DEVISE_PEPPER=`openssl rand -base64 40` \
-  DEVISE_SECRET_KEY=`openssl rand -base64 40` \
-  GOVUK_WEBSITE_ROOT=https://www.gov.uk \
-  bundle exec rails assets:clean assets:precompile
-
-HEALTHCHECK CMD curl --silent --fail localhost:$PORT || exit 1
-
+ARG base_image=ruby:2.7.6-slim-bullseye
+
+FROM $base_image AS builder
+ENV RAILS_ENV=production \
+    NODE_ENV=production \
+    GOVUK_APP_DOMAIN=www.gov.uk \
+    GOVUK_WEBSITE_ROOT=https://www.gov.uk \
+    ASSETS_PREFIX=/assets/signon \
+    BOOTSNAP_CACHE_DIR=/var/cache/bootsnap
+# TODO: have a separate build image which already contains the build-only deps.
+RUN apt-get update -qy && \
+    apt-get upgrade -y && \
+    apt-get clean
+RUN apt-get install -y build-essential nodejs gnupg2 curl libmariadb-dev-compat
+RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add - && \
+    echo "deb https://dl.yarnpkg.com/debian/ stable main" > /etc/apt/sources.list.d/yarn.list && \
+    apt-get install -y yarn
+
+RUN mkdir /app
+WORKDIR /app
+COPY Gemfile Gemfile.lock .ruby-version /app/
+RUN bundle config set without 'development test' && \
```

Comment:
I looked into this in quite some detail and concluded that omitting it was probably the lesser of evils, since it messes with where it installs the gems (because it was designed for a pre-container era). The deciding factor for me was that I really want to avoid the surprise/confusion of having gems installed in more than one place in the image filesystem, since the whole container is inherently specific to a single app. 

I can only see one potential upside to enabling this option: it might detect a scenario where someone has merged a change to `Gemfile` but not `Gemfile.lock` and the tests have nevertheless passed. Personally I think if we're concerned about that scenario then it'd be much kinder to the user if we detect it before merge. (I think the nicest way would be to add a pre-merge check that `bundle update` on a fresh checkout produces no diff.)

If you're still keen on enabling the option, I'd be open to that as long as it still installs the gems in the location that we've configured, i.e. doesn't introduce a second location that's different from the system-wide one. (I'd prefer to write a pre-merge check for Gemfile/Gemfile.lock consistency though!)

Reply:
Oh just realised that might be a bit confusing cos I think I haven't raised the PR that deals with the GEM_HOME/BUNDLE_PATH stuff yet 😅 I'll keep you posted.